pkg/apis/resource/validation/validation.go

```diff
@@ -214,146 +302,130 @@ func ValidateClaimStatusUpdate(resourceClaim, oldClaim *resource.ResourceClaim)
-			allErrs = append(allErrs, validateItem(item, idxPath)...)
-			allItems.Insert(item)
-		}
-	}
-	if len(slice) > maxSize {
-		// Dumping the entire field into the error message is likely to be too long,
-		// in particular when it is already beyond the maximum size. Instead this
-		// just shows the number of entries.
-		allErrs = append(allErrs, field.TooLongMaxLength(fldPath, len(slice), maxSize))
-	}
+// ValidateClass validates a DeviceClass.
+func ValidateDeviceClass(class *resource.DeviceClass) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMeta(&class.ObjectMeta, false, corevalidation.ValidateClassName, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateDeviceClassSpec(&class.Spec, nil, field.NewPath("spec"))...)
+	return allErrs
+}
+
+// ValidateClassUpdate tests if an update to DeviceClass is valid.
+func ValidateDeviceClassUpdate(class, oldClass *resource.DeviceClass) field.ErrorList {
+	allErrs := corevalidation.ValidateObjectMetaUpdate(&class.ObjectMeta, &oldClass.ObjectMeta, field.NewPath("metadata"))
+	allErrs = append(allErrs, validateDeviceClassSpec(&class.Spec, &oldClass.Spec, field.NewPath("spec"))...)
 	return allErrs
 }
 
-// validateResourceClaimConsumers ensures that the slice contains no duplicate UIDs and does not exceed a certain maximum size.
-func validateResourceClaimConsumers(consumers []resource.ResourceClaimConsumerReference, maxSize int, fldPath *field.Path) field.ErrorList {
+func validateDeviceClassSpec(spec, oldSpec *resource.DeviceClassSpec, fldPath *field.Path) field.ErrorList {
 	var allErrs field.ErrorList
-	allUIDs := sets.New[types.UID]()
-	for i, consumer := range consumers {
-		idxPath := fldPath.Index(i)
-		if allUIDs.Has(consumer.UID) {
-			allErrs = append(allErrs, field.Duplicate(idxPath.Child("uid"), consumer.UID))
-		} else {
-			allErrs = append(allErrs, validateResourceClaimUserReference(consumer, idxPath)...)
-			allUIDs.Insert(consumer.UID)
-		}
+	// If the selectors are exactly as before, we treat the CEL expressions as "stored".
+	// Any change, including merely reordering selectors, triggers validation as new
+	// expressions.
+	stored := false
+	if oldSpec != nil {
+		stored = apiequality.Semantic.DeepEqual(spec.Selectors, oldSpec.Selectors)
 	}
-	if len(consumers) > maxSize {
-		// Dumping the entire field into the error message is likely to be too long,
-		// in particular when it is already beyond the maximum size. Instead this
-		// just shows the number of entries.
-		allErrs = append(allErrs, field.TooLongMaxLength(fldPath, len(consumers), maxSize))
+	allErrs = append(allErrs, validateSlice(spec.Selectors, resource.DeviceSelectorsMaxSize,
+		func(selector resource.DeviceSelector, fldPath *field.Path) field.ErrorList {
+			return validateSelector(selector, fldPath, stored)
+		},
+		fldPath.Child("selectors"))...)
+	allErrs = append(allErrs, validateSlice(spec.Config, resource.DeviceConfigMaxSize, validateDeviceClassConfiguration, fldPath.Child("config"))...)
```

Comment:
Triggered by the "why re-validate unchanged fields" question I am starting to wonder about the common (right?) practice of calling the normal validate function for create as part of an update, like I am doing here.

`spec.Config` contains a one-of. We check that `Opaque` is set. Suppose the following:
- In 1.32, we add an alternative.
- Someone creates a DeviceClass with a config using that alternative.
- They downgrade to 1.31.
- The DeviceClass is now invalid (empty one-of).

What operations should be allowed in that state?

An update which removes the config works, the updated DeviceClass is valid again.

Reading the DeviceClass works. Clients bulk at processing it because they check the one-of. Also okay.

What about deleting the DeviceClass? I suppose if it can be removed directly, it's okay. But if it has a finalizer, the DeviceClass must be updated (add DeletionTimestamp), which probably fails.

Thinking this through further, this might be the reason why "unchanged" fields should be re-validated: if the current apiserver does not support some future field, then any update will drop that field because the apiserver doesn't know how to re-encode it. Validation is not guaranteed to catch this, but it does for one-ofs and enums. So if an "apparently" unmodified field no longer passes validation, it's likely that the apiserver lacks something and it might be better to not update the resource because that would drop information.

Reply:
You left out feature gates.

* `spec.Config` contains a one-of. We check that `Opaque` is set.
* In 1.32 we move DRA to beta (on by default)
* In 1.32, we add an alternative config under a new gate, which is alpha
* Someone enables that alpha gate and creates a DeviceClass with a config using that alternative.
* They downgrade to 1.31.
* The DeviceClass is now invalid (empty one-of).

There's no rollback-safety when using alpha gates.  If instead:

* `spec.Config` contains a one-of. We check that `Opaque` is set.
* In 1.32 we move DRA to beta (on by default)
* In 1.32, we add an alternative config under a new gate, which is alpha
* In 1.33 we move that new gate to beta
* Someone creates a DeviceClass with a config using that alternative.
* They downgrade to 1.32.
* The DeviceClass is now valid because of the "already in use" logic

We COULD say that one-ofs are allowed to have no value set, and consumers should ignore them, but we usually don't because it makes edge-cases that don't need to exist.